include/oneapi/dpl/experimental/kt/internal/esimd_radix_sort_kernels.h

```diff
@@ -0,0 +1,688 @@
+    {
+        __dpl_esimd_ns::fence<__dpl_esimd_ns::fence_mask::sw_barrier>();
+        __dpl_esimd_ns::simd<::std::uint32_t, 32> __matched_bins(0xffffffff);
+        _ONEDPL_PRAGMA_UNROLL
+        for (int __i = 0; __i < __radix_bits; __i++)
+        {
+            __dpl_esimd_ns::simd<::std::uint32_t, 32> __bit = (__bins >> __i) & 1;
+            __dpl_esimd_ns::simd<::std::uint32_t, 32> __x =
+                __dpl_esimd_ns::merge<::std::uint32_t, 32>(0, -1, __bit != 0);
+            ::std::uint32_t __ones = __dpl_esimd_ns::pack_mask(__bit != 0);
+            __matched_bins = __matched_bins & (__x ^ __ones);
+        }
+        __dpl_esimd_ns::fence<__dpl_esimd_ns::fence_mask::sw_barrier>();
+        return __matched_bins;
+    }
+
+    inline auto
+    __rank_local(_LocOffsetSimdT& __ranks, _LocOffsetSimdT& __bins,
+                 ::std::uint32_t __slm_counter_offset, ::std::uint32_t __local_tid) const
+    {
+        constexpr int __bins_per_step = 32;
+        using _ScanSimdT = __dpl_esimd_ns::simd<::std::uint32_t, __bins_per_step>;
+
+        __utils::__block_store_slm<_LocOffsetT, __bin_count>(__slm_counter_offset, 0);
+        _ScanSimdT __remove_right_lanes, __lane_id(0, 1);
+        __remove_right_lanes = 0x7fffffff >> (__bins_per_step - 1 - __lane_id);
+
+        static_assert(__data_per_work_item % __bins_per_step == 0);
+        _ONEDPL_PRAGMA_UNROLL
+        for (::std::uint32_t __s = 0; __s < __data_per_work_item; __s += __bins_per_step)
+        {
+            _ScanSimdT __this_bins = __bins.template select<__bins_per_step, 1>(__s);
+            _ScanSimdT __matched_bins = __match_bins(__this_bins, __local_tid);
+            _ScanSimdT __pre_rank = __utils::__vector_load<_LocOffsetT, 1, __bins_per_step>(
+                __slm_counter_offset + __this_bins * sizeof(_LocOffsetT));
+            auto __matched_left_lanes = __matched_bins & __remove_right_lanes;
+            _ScanSimdT __this_round_rank = __dpl_esimd_ns::cbit(__matched_left_lanes);
+            auto __this_round_count = __dpl_esimd_ns::cbit(__matched_bins);
+            auto __rank_after = __pre_rank + __this_round_rank;
+            auto __is_leader = __this_round_rank == __this_round_count - 1;
+            __utils::__vector_store<_LocOffsetT, 1, __bins_per_step>(
+                __slm_counter_offset + __this_bins * sizeof(_LocOffsetT), __rank_after + 1, __is_leader);
+            __ranks.template select<__bins_per_step, 1>(__s) = __rank_after;
+        }
+        __dpl_esimd_ns::barrier();
+    }
+
+    inline void
+    __rank_global(_LocHistT& __subgroup_offset, _GlobHistT& __global_fix,
+                  ::std::uint32_t __local_tid, ::std::uint32_t __wg_id) const
+    {
+        /*
+        first do column scan by group, each thread do 32c,
+        then last row do exclusive scan as group incoming __offset
+        then every thread add local sum with sum of previous group and incoming __offset
+        */
```

Comment:
Similar comment as before regarding moving the comments to their respective code sections.

Reply:
I have described this stage more thoroughly, each its step individually.